modules/regexp-parser/regexp-parser.c

```diff
@@ -0,0 +1,170 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "regexp-parser.h"
+#include "parser/parser-expr.h"
+#include "scratch-buffers.h"
+#include "string-list.h"
+
+#include <string.h>
+#include <stdio.h>
+
+typedef struct _RegexpParser
+{
+  LogParser super;
+  gchar *prefix;
+  GList *patterns;
+  LogMatcherOptions matcher_options;
+  GList *matchers;
+} RegexpParser;
+
+LogMatcherOptions *
+regexp_parser_get_matcher_options(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  return &self->matcher_options;
+}
+
+void
+regexp_parser_set_prefix(LogParser *s, const gchar *prefix)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  g_free(self->prefix);
+  if (prefix)
+    self->prefix = g_strdup(prefix);
+  else
+    self->prefix = NULL;
+}
+
+void
+regexp_parser_set_patterns(LogParser *s, GList *patterns)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  string_list_free(self->patterns);
+  self->patterns = patterns;
+}
+
+gboolean
+regexp_parser_compile(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  log_matcher_options_init(&self->matcher_options);
+  for (GList *item = self->patterns; item; item = item->next)
+    {
+      self->matchers = g_list_prepend(self->matchers, log_matcher_new(&self->matcher_options));
+      log_matcher_pcre_set_prefix((LogMatcher *)self->matchers->data, self->prefix);
+
+      if(!log_matcher_compile((LogMatcher *)self->matchers->data, item->data, NULL))
+        return FALSE;
+    }
+
+  self->matchers = g_list_reverse(self->matchers);
+  return TRUE;
+}
+
+static gboolean
+regexp_parser_process(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input,
+                      gsize input_len)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  log_msg_make_writable(pmsg, path_options);
+  msg_trace("regexp-parser message processing started",
+            evt_tag_str ("input", input),
+            evt_tag_str ("prefix", self->prefix),
+            evt_tag_printf("msg", "%p", *pmsg));
+
+  gboolean result = FALSE;
+  for (GList *item = self->matchers; item; item = item->next)
+    {
+      msg_trace("regexp-parser message processing for",
+                evt_tag_str ("input", input),
+                evt_tag_str ("pattern", ((LogMatcher *)item->data)->pattern));
+
+      if (log_matcher_match((LogMatcher *)item->data, *pmsg, LM_V_NONE, input, input_len))
```

Comment:
although it's a bit of a layering violation, we may actually know that we are matching against LM_V_MESSAGE (in case the template() option is unset). And with that knowledge the LogMatcher instance can store indirect matches instead of full copies.

So instead of LM_V_NONE, you could use a conditional to see if ```self->super.template``` is set, and use LM_V_NONE if it is, and LM_V_MESSAGE if it's not.

This would result in the use of "indirect references" within the LogMessage, meaning that slices of the MESSAGE payload would be used as values. This might compress the size of the serialized log message and also reduce CPU usage as the new name-value pair is stored.